src/dune_engine/sandbox.ml

```diff
@@ -69,8 +77,18 @@ let link_function ~(mode : Sandbox_mode.some) =
       | true -> win32_error mode
       | false -> fun src dst -> Io.portable_hardlink ~src ~dst))
 
-let link_deps t ~mode ~deps =
-  let link = Staged.unstage (link_function ~mode) in
+let copy_and_make_writable src dst =
+  Io.copy_file ~src ~dst ~chmod:(fun n -> n lor 0o200) ()
+
+let link_deps t ~mode ~patch_back_source_tree ~deps =
+  let link =
+    if patch_back_source_tree then
```

Comment:
Silently overriding the mode chosen by the user doesn't seem great. Or, perhaps, we can only reach this point if `mode = Copy`, because otherwise we'd fail somewhere earlier? If this is the case, then I think we should make this override in `link_function` (specifically, in the `Copy` case).

Reply:
AFAICT, the user cannot request the symlink or hardlink mode via a `dune` file. All one can specify in a `dune` file is to:

- never sandbox a rule
- always sandbox a rule
- always sandbox a rule, preserving file kinds, which forbids symlink mode

So here we can never override anything meaningful.

The only way to request a particular mode is via the command line. In this case, we simply ignore the command line setting, which seems fair for this kind of rules. I added a test for this as well.

Reply:
Right, if we can't currently override this on a per-action basis, then it's probably OK to occasionally fall back to copying if the global setting is `Symlink` or `Hardlink`. It still seems a bit fishy, but I don't see a better option.

Reply:
Same. And again, it's for pretty dodgy rules. In the long term, I expect that we'll provide enough good features for users to move away from this. So I think it's ok if this feature has rough edges. What people currently do is already problematic, so we are not making things that much worse.